### DISCUSSION TITLE
Modules: Always autoload run dependencies

### DISCUSSION THREAD

**Discussion by adamjstewart** [spack/spack#8639] [Ideas]
One of the tenets of Spack is the use of RPATH. It allows packages to find a complicated set of dependencies and prevents users from loading incompatible dependencies. Unfortunately, RPATH doesn't extend to non-compiled languages. Python, Perl, R, and Java all require their dependencies to be found via `PYTHONPATH`, `PERL5LIB`, `R_LIBS`, and `CLASSPATH`, respectively. As a package manager, Spack is great at managing dependencies. This should extend to its auto-generated module files.

### Expected Result

If I need to use a software package, I should be able to `module load` that package and go.

### Actual Result

For many packages, I also need to `module load` all of their runtime dependencies. For example, if I want to build our docs, I need to:
```console
$ module load py-sphinx
$ module load py-alabaster
$ module load py-setuptools
$ module load py-babel
$ module load py-pytz
$ module load py-docutils
$ module load py-imagesize
$ module load py-jinja2
$ module load py-markupsafe
$ module load py-packaging
$ module load py-pyparsing
$ module load py-six
$ module load py-pygments
$ module load py-requests
$ module load py-snowballstemmer
$ module load py-sphinx-rtd-theme
$ module load py-sphinxcontrib-websupport
$ module load py-typing
$ module load python
```
The situation for `py-matplotlib` is even worse. And then there's `r-rminer`...

### Existing Alternatives

There are some ways to get around this issue. Spack already has builtin support for autoloading dependencies: http://spack.readthedocs.io/en/latest/module_file_support.html#autoload-dependencies

However, this requires manual intervention on the part of the Spack admin. It also needs to be done for every non-compiled programming language. There are also many compiled programs that pick up runtime dependencies via `PATH`. Although they are listed as `type='run'` dependencies in the package, this information isn't used anywhere during modulefile generation (as far as I know).

One can also `spack activate` extendable packages like Python and R. Again, this only works for extendable packages; there are many non-extendable packages that suffer from the same problem.

`spack view` is another alternative, but not always possible when multiple configurations of the same package are installed.

One can `spack load -r` to recursively load dependencies, but this will load many unnecessary dependencies when the only thing you actually need are 'run' dependencies.

Many packages have taken matters into their own hands. See #8629 for one of dozens of packages that manually put their 'run' dependencies in the `PATH`.

### Proposed Solutions

I see 2 possible solutions to this dilemma. Either one of these should be enabled by default (or even non-negotiable). This will greatly reduce the difficulty that new users face when starting to use Spack.

1. Autoload modules for all 'run' dependencies. The modulefile itself would load the modules for its dependencies. (Does this work for all 3 module environments that Spack supports?)
2. Manually set certain environment variables to include 'run' dependencies. This would include `PATH`, `PYTHONPATH`, etc.

Can anyone think of arguments for or against 1 or 2? Is there any reason Spack shouldn't always autoload 'run' dependencies?

@tgamblin @alalazo @hartzell @healther @matthiasdiener

**Comment by matthiasdiener** [comment]
I'm voting for solution (1). Otherwise, what would be the point of having a "run" dependency?

Also, I think the current manual is somewhat misleading on this situation. Specifically, 
http://spack.readthedocs.io/en/latest/packaging_guide.html#dependency-types states that:

> “run”: the project is used by the project at runtime. The package will be added to PATH and PYTHONPATH

**Comment by healther** [comment]
Strongly agree with (1), I asked that question some time ago, but it didn't gain any traction (and I think there even was a reason given for why not, but I don't remember and probably didn't understand it completely). Being able to use `spack load <package>` instead of `spack load -r <package>` is imo more natural

**Comment by alalazo** [comment]
> Although they are listed as type='run' dependencies in the package, this information isn't used anywhere during modulefile generation (as far as I know).

That's correct. Branches of different types (`build`, `run`, `link`, `tests`) were added after module support, and module generation has never been updated with respect to that. So I think you're right: it's time we add something that makes good use of that information when generating modules. I'll try to come up with a proposal (and push in the meanwhile for #7098 to go forward).

**Comment by alalazo** [comment]
@healther @matthiasdiener Autoloading comes with its own drawbacks. This is just to say that autoloading  might be the default, if people agree with that, but certainly not the only choice. I think module generation should remain highly configurable as sites that use modules seem to be very opinionated on how they want to deploy them.

**Comment by healther** [comment]
@alalazo: As we introduced `spack view` that solved my problem I haven't looked to hard into modules anymore. Take my comments more as a "I don't understand why someone wants it that way" rather than a "wanting it that way is stupid" :)

**Comment by matthiasdiener** [comment]
As an aside, `spack load -r` does not seem to be documented at all. https://spack.readthedocs.io/en/latest/command_index.html#spack-load only mentions an `-h|--help` option, which actually only forwards this option to the module command.

**Comment by healther** [comment]
yeah, we could add a help page, but if someone would be willing to put in the work to actually do a clean implementation that would be better. I'll probably have some time on Friday, so if no one has shouted here I'll do the documentation :)

**Comment by hartzell** [comment]
For what it's worth, here is the Lmod documentation for [handling dependencies](http://lmod.readthedocs.io/en/latest/098_dependent_modules.html).

It seems like our default Lmod configuration should minimally include `prereq(...)` statements for the run-time dependencies, so that Lmod generates useful error messages instead of allowing the user to experience run-time failures.

There are two other choices: `load()` and `always_load()`.  Modules loaded with `load()` are unloaded if/when the dependee is unloaded.  Modules loaded with `always_load()` are *not* unloaded if/when the dependee is unloaded.

In an ideal world, I think that what I would want is to unload the dependent if it wasn't loaded explicitly and is no longer used by other modulefiles.

**Comment by adamjstewart** [comment]
> In an ideal world, I think that what I would want is to unload the dependent if it wasn't loaded explicitly and is no longer used by other modulefiles.

Agreed. Of the options listed, I think `depends_on()` is the closest to the behavior I would expect.

**Comment by hartzell** [comment]
[edit, typo]

I made a quick hack to the `modulefile.lua` template file so that it uses `depends_on` instead of `load`.

```diff
diff --git a/templates/modules/modulefile.lua b/templates/modules/modulefile.lua
index cf37595..9b19db5 100644
--- a/templates/modules/modulefile.lua
+++ b/templates/modules/modulefile.lua
@@ -61,12 +61,7 @@ setenv("LMOD_{{ name|upper() }}_VERSION", "{{ version_part }}")

 {% block autoloads %}
 {% for module in autoload %}
-if not isloaded("{{ module }}") then
-{% if verbose %}
-    LmodMessage("Autoloading {{ module }}")
-{% endif %}
-    load("{{ module }}")
-end
+depends_on("{{ module }}")
 {% endfor %}
 {% endblock %}
```

It seems to work quite nicely.  Both `gcc` and `curl` depend on `zlib` and you can see in the example below that it's not unloaded until it's safe.

```
[hartzell@ip-10-96-1-107 spack]$ module purge
[hartzell@ip-10-96-1-107 spack]$ module load gcc
[hartzell@ip-10-96-1-107 spack]$ module list

Currently Loaded Modules:
  1) gcc/5.5.0-26dw7wa



[hartzell@ip-10-96-1-107 spack]$ module load emacs
[hartzell@ip-10-96-1-107 spack]$ module list

Currently Loaded Modules:
  1) gcc/5.5.0-26dw7wa     4) nettle/3.3-gm37wru    7) libxml2/2.9.8-hgfwbir  10) gettext/0.19.8.1-vf4khxi  13) emacs/26.1-nyfkrps
  2) ncurses/6.1-b3yjgmv   5) zlib/1.2.11-56rgaos   8) bzip2/1.0.6-cwlahss    11) gnutls/3.5.13-4oa5gq7
  3) gmp/6.1.2-cd47nos     6) xz/5.2.4-xz3h7mc      9) tar/1.30-ah7apno       12) pcre/8.42-idvaazk



[hartzell@ip-10-96-1-107 spack]$ module load curl
[hartzell@ip-10-96-1-107 spack]$ module list

Currently Loaded Modules:
  1) gcc/5.5.0-26dw7wa     4) nettle/3.3-gm37wru    7) libxml2/2.9.8-hgfwbir  10) gettext/0.19.8.1-vf4khxi  13) emacs/26.1-nyfkrps
  2) ncurses/6.1-b3yjgmv   5) zlib/1.2.11-56rgaos   8) bzip2/1.0.6-cwlahss    11) gnutls/3.5.13-4oa5gq7     14) openssl/1.0.2n-66jppsd
  3) gmp/6.1.2-cd47nos     6) xz/5.2.4-xz3h7mc      9) tar/1.30-ah7apno       12) pcre/8.42-idvaazk         15) curl/7.60.0-mnkg6zw



[hartzell@ip-10-96-1-107 spack]$ module load jq
[hartzell@ip-10-96-1-107 spack]$ module list

Currently Loaded Modules:
  1) gcc/5.5.0-26dw7wa     4) nettle/3.3-gm37wru    7) libxml2/2.9.8-hgfwbir  10) gettext/0.19.8.1-vf4khxi  13) emacs/26.1-nyfkrps      16) oniguruma/6.1.3-mo2fout
  2) ncurses/6.1-b3yjgmv   5) zlib/1.2.11-56rgaos   8) bzip2/1.0.6-cwlahss    11) gnutls/3.5.13-4oa5gq7     14) openssl/1.0.2n-66jppsd  17) jq/1.5-bic4vte
  3) gmp/6.1.2-cd47nos     6) xz/5.2.4-xz3h7mc      9) tar/1.30-ah7apno       12) pcre/8.42-idvaazk         15) curl/7.60.0-mnkg6zw



[hartzell@ip-10-96-1-107 spack]$ module unload emacs
[hartzell@ip-10-96-1-107 spack]$ module list

Currently Loaded Modules:
  1) gcc/5.5.0-26dw7wa   2) zlib/1.2.11-56rgaos   3) openssl/1.0.2n-66jppsd   4) curl/7.60.0-mnkg6zw   5) oniguruma/6.1.3-mo2fout   6) jq/1.5-bic4vte



[hartzell@ip-10-96-1-107 spack]$ module unload curl
[hartzell@ip-10-96-1-107 spack]$ module list

Currently Loaded Modules:
  1) gcc/5.5.0-26dw7wa   2) oniguruma/6.1.3-mo2fout   3) jq/1.5-bic4vte



[hartzell@ip-10-96-1-107 spack]$ module unload jq
[hartzell@ip-10-96-1-107 spack]$ module list

Currently Loaded Modules:
  1) gcc/5.5.0-26dw7wa



[hartzell@ip-10-96-1-107 spack]$
```

@alalazo 
- What do you think of switching the default template from using `load()` to using `depends_on()`?
- Is there a way to have a "local" copy of the template (like we do for the various YAML files) or would I just ~override~ overwrite  it?

**Comment by adamjstewart** [comment]
> Is there a way to have a "local" copy of the template (like we do for the various YAML files) or would I just override it?

Yes, @alalazo added logic so that you can override it. See http://spack.readthedocs.io/en/latest/tutorial_modules.html#working-with-templates

**Comment by hartzell** [comment]
> @alalazo added logic

**Brilliant**!

**Comment by citibeth** [comment]
@adamjstewart Pleaser review #2982 for a previous discussion of this issue.

**Comment by adamjstewart** [comment]
@citibeth I think a lot has changed since #2982. Now that `depends_on` has a `type='run'` argument, we now have a way of directly knowing which packages need to be recursively loaded.

**Comment by citibeth** [comment]
Thanks for reading and digesting it!

**Comment by tgamblin** [comment]
I think `depends_on` is the right thing to do for `lmod`.  What's the right thing for `tcl` modules?  Is there an equivalent?  AFAIK they do not do reference counting in the environment the way `lmod` does.

**Comment by tgamblin** [comment]
I think #2982 was the right idea and agree with @adamjstewart's great description here that we should autoload run dependencies by default.  It looks like #2982 was aimed at implementing a lot of this stuff *in* spack, but really as @adamjstewart says, the module systems should handle this based on the dependency relationships Spack already has.

To be honest, I assumed this was easy to configure and that we had this right in the modules implementation.  Spack was designed with compiled packages in mind, so we go to great lengths to get `RPATH` right.  So, sorry for not looking more in-depth at this.  Loading what is needed for packages to run should be the default and shouldn't require configuration.

**Comment by tgamblin** [comment]
Also, FYI: @becker33 and I visited @alalazo at EPFL week before last and we set up a project to improve modules support.  This issue is in it.  More [here](https://github.com/spack/spack/projects/4).

**Comment by citibeth** [comment]
Lower-level modules are used many times in a typical DAG.  How do we ensure
they are not loaded many times with recursive loading?

**Comment by tgamblin** [comment]
@citibeth: `lmod` is reference counted (see docs above).  I assume that `tcl` modules do something similar but I am not sure.  Both store state in the environment.

**Comment by citibeth** [comment]
I'm not sure either, maybe @alalazo knows?

In any case... what I REALLY want is Spack Environments to generate a
single module that loads the entire environment.  That would be efficient
especially on supercomputers with slow FS.  And it would eliminate all
those hashes from anything user-facing.

But first I'm waiting for Spack Environments to get merged....??????

.. Elizabeth

**Comment by tgamblin** [comment]
It's going to happen :)

**Comment by alalazo** [comment]
@hartzell
>What do you think of switching the default template from using load() to using depends_on()? 

I am looking into that right now. It seems fine to me at a first glance. There's a known caveat:

> When gcc is unloaded it removes the path in MODULEPATH to the gcc dependent modules which means that X will be unloaded and marked as inactive. The way that a module is unloaded is that the contents of the module is evaluated and most action requested are reversed. So load statements cause a module to unload and a depends_on() function is told to forgo() the modules. The isloaded() is not reversed. But as you can see since the gcc modulefile is not loaded the if statement then clause is not evaluated. This means that openblas will still be loaded.

but I don't think we'll ever hit that (*famous last words*). I need to check since which version of `LMod` is `depends_on` supported, but in the worst case we can let a user select which idiom to use for autoloading `lmod` modules in `modules.yaml`. What do you think?

**Comment by hartzell** [comment]
@alalazo -- Works for me.  I don't think [*hey, I'm famous too...* :)] that caveat will be a problem for me in my current use-case (aka "job").   Others (@peetsv) might be able to speak to the old use-case (aka "previous gig") but I don't think it'll be a problem for them either.

**Comment by ferdonline** [comment]
Are there any news here? 
I agree with the argument that loading modules for every dependency might kill performance in many applications especially HPC, as @citibeth just mentioned. But I also believe this is mostly problematic with extensions (like Py packages) which are loaded by the dozens. Eventually in the case of pure dependencies we are just fine recursively loading modules.

Maybe it wouldn't be such a bad idea to only optimize the case of extensions (to which many sites don't like creating modules), so their environment setup would always be "merged" into the 'requiring' software environment.

**Comment by bebosudo** [comment]
Hi! Any update regarding this issue?

**Comment by citibeth** [comment]
Yes, spack environments are now merged.  The command ‘spack env loads -r’
will generate a shell script that will load the complete list of modules
you need including dependencies.

**Comment by bebosudo** [comment]
Hi @citibeth, thanks for your reply, but I was referring to the "autoload" of dependencies of type "run".
In particular, I tried adding the "run" type to regcm as requested also by issue #10163, but spack didn't add the `/bin/` directory of the dependency to the module generated for regcm, see discussion with Adam in PR #10863.

**Comment by adamjstewart** [comment]
Yes, spack environments are a great tool for single users, but for sys admins who want to deploy a set of modules that "Just Work", we still have a long ways to go with autoloading run deps.

**Comment by citibeth** [comment]
Spack environments are exactly the tool we have available to describe a set
of modules that will “just work”.  As Todd has pointed out, manually loaded
modules are fundamentally broken.  No amount of auto loading will fix this.

**Reply by haampie** [reply]
> As Todd has pointed out, manually loaded modules are fundamentally broken.

I don't get that from this discussion, especially considering

> I think `depends_on` is the right thing to do for `lmod`

it can be done properly

**Comment by haampie** [comment]
As an aside to the module discussion:

Let me point out that the `-r` flag in `spack load -r` has *no effect whatsoever*, it's an unused flag, probably an oversight. See https://github.com/spack/spack/pull/25731. The current default of `spack load x` is to load `x + all dependencies`, including build deps of build deps of build deps, which doesn't make sense.

And then if https://github.com/spack/spack/pull/25732 gets merged too, `spack load x` will do exactly what this thread wants `module load x` to do, namely load `x + runtime deps`. This would then be similar to `spack env activate x`, which only loads `root specs of env x + runtime deps`.
